forward/forwarder_test.go

```diff
@@ -65,17 +61,43 @@ type Pong struct {
 }
 
 func (s *ForwarderTestSuite) registerPong(address string, channel *tchannel.Channel) {
-	handler := func(ctx json.Context, ping *Ping) (*Pong, error) {
-		return &Pong{"Hello, world!", address}, nil
+	hmap := map[string]interface{}{
+		"/ping": func(ctx json.Context, ping *Ping) (*Pong, error) {
+			return &Pong{"Hello, world!", address}, nil
+		},
+		"/error": func(ctx json.Context, ping *Ping) (*Pong, error) {
+			return nil, errors.New("remote error")
+		},
 	}
-	hmap := map[string]interface{}{"/ping": handler}
+	s.Require().NoError(json.Register(channel, hmap, func(ctx context.Context, err error) {}))
+
+	thriftHandler := &pingpong.MockTChanPingPong{}
 
-	s.Require().NoError(json.Register(channel, hmap,
-		func(ctx context.Context, err error) {}))
+	// successful request
+	thriftHandler.On("Ping", mock.Anything, &pingpong.Ping{
+		Key: "success",
+	}).Return(&pingpong.Pong{
+		Source: address,
+	}, nil)
+
+	// error request
+	thriftHandler.On("Ping", mock.Anything, &pingpong.Ping{
+		Key: "error",
+	}).Return(nil, &pingpong.PingError{})
+
+	server := thrift.NewServer(channel)
+	server.Register(pingpong.NewTChanPingPongServer(thriftHandler))
 }
 
 func (s *ForwarderTestSuite) SetupSuite() {
-	s.sender = &DummySender{"127.0.0.1:3001", "127.0.0.1:3001"}
+	sender := &MockSender{}
+	sender.On("Lookup", "me").Return("127.0.0.1:3001", nil)
```

Comment:
Don't really know what you mean with port 0 and collisions. These lines are hardcoded responses to lookup for specific keys. So no actual listening is happening here. And most of the ports are not being listened to. Only port 3002 has a listener iirc. The listener for this port is created on line 109.

I don't know what the implications of port 0 would be. If a ring is used it is deterministic now since the ip:port's of all nodes are hardcoded.